framework/src/modules/chain/synchronizer/synchronizer.js

```diff
@@ -1,44 +1,162 @@
 const BlockSynchronizationMechanism = require('./block_synchronization_mechanism');
 const FastChainSwitchingMechanism = require('./fast_chain_switching_mechanism');
```

Comment:
As discussed, in my opinion, these dependencies should be instantiated in `Chain.js` and registered into Synchronizer there. Doing the way it is right now will make Synchronizer dependencies grow if one of the mechanisms needs more of them.

Reply:
If we initiate these in the chain, we have to register logic dynamically when to choose which strategy. Which I feel is not bad idea either, but considering that for reasonable time in future we are not going to introduce new sync mechanism, this approach is fine for now.

Reply:
I understand, however, the purpose of having a Synchronizer on top was to isolate implementation details from other parts of the code. For me even if it is not fully complete, it makes more sense to:

1. Instantiate both mechanism classes with its corresponding dependencies
2. Register them to the Synchronizer, so the Synchronizer shouldn't have to include the dependencies of the mechanisms registered in it. If in the future (possibly), we improve or change a mechanism, we might have to add/remove dependencies and that will mean we will have to modify the upper Synchronizer class, this will make the purpose of the Synchronizer empty.

IMHO it makes more sense to do it this way since the beginning than have it to refactor it later on.

Reply:
As said earlier, If I see even smaller percent of chance to have new synchronisation mechanism or any SDK feature to register custom sync mechanism, I would have gone with the direction you suggested. 

But for now we have only two and we keep go with these two and not exposing it as feature for the SDK to register custom sync mechanism. @MaciejBaj Can suggest if we plan for any such feature for future. 

For suggested above reasons, I think with current approach. If we plan for such feature for SDK, then we will brainstorm further on how to expose the sync selection logic for dynamic use.

Reply:
IMO, it’s not meant to be exposed to the higher layer of the SDK for now but to ease future development on both mechanisms and to simplify its design.

As I said, doing it the way you propose will induce extra development if any of the dependencies of each mechanism change in the future (which we don’t know, and doesn’t necessarily mean we add more mechanisms) and Synchronizer will have to be modified. In my view, doing it the way I propose since the beginning can save us time in the future and the difference in implementation is very simple, just registering the mechanisms to the Synchronizer in Chain.js.

I find more valuable to go the way I propose for the reasons I stated than having to revisit this in the future.

Reply:
I don't see the difference of maintaining effort for both approaches. If there is a change, you have to maintain equally in both cases. 

Will see if I can came up with a middle ground.